File: scripts/verify_sdk_parity.py
import json
import sys
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
EXPECTED = 158


def load_bindings(package_src: Path, module_name: str) -> list[dict]:
    sys.path.insert(0, str(package_src))
    try:
        module = __import__(f"{module_name}._mcp_generated_client", fromlist=["MCP_TOOL_BINDINGS"])
        return list(module.MCP_TOOL_BINDINGS)
    finally:
        sys.path.pop(0)
        for loaded in list(sys.modules):
            if loaded == module_name or loaded.startswith(f"{module_name}."):
                del sys.modules[loaded]


def to_snake_case(value: str) -> str:
    value = re.sub(r"[-\s]+", "_", value)
    return re.sub(r"(?<!^)(?=[A-Z])", "_", value).lower().strip("_")


def assert_exact(label: str, actual: list[dict], expected: list[dict]) -> None:
    actual_names = {binding["name"] for binding in actual}
    expected_names = {binding["name"] for binding in expected}
    missing = sorted(expected_names - actual_names)
    extra = sorted(actual_names - expected_names)
    expected_by_name = {
        tool["name"]: (tool["category"], to_snake_case(tool["methodName"]))
        for tool in expected
    }
    drift = sorted(
        binding["name"]
        for binding in actual
        if expected_by_name.get(binding["name"]) != (binding["category"], binding["method_name"])
    )
    if len(actual) != EXPECTED or missing or extra or drift:
        raise RuntimeError(
            f"{label} parity failed: count={len(actual)}, missing={missing}, extra={extra}, binding_drift={drift}"
        )
    print(f"{label}: {len(actual)}/{EXPECTED}")


def main() -> None:
    manifest = json.loads((ROOT / "contracts" / "mcp.tools.json").read_text())
    expected = manifest["tools"]
    assert_exact(
        "Python scraper package",
        load_bindings(ROOT / "packages" / "scraper-python" / "src", "mcpscraper"),
        expected,
    )
    assert_exact(
        "Python memory package",
        load_bindings(ROOT / "packages" / "memory-python" / "src", "mcpscraper_memory"),
        expected,
    )


if __name__ == "__main__":
    main()

File: scripts/verify-sdk-parity.ts
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import * as memorySdk from '../packages/memory/src/index.js'
import * as scraperSdk from '../packages/scraper/src/index.js'
import { MCP_TOOL_CATALOG } from '../packages/cli/src/generated-tools.js'

const EXPECTED = 159

function namesFromBindings(bindings: ReadonlyArray<{ name: string }>): string[] {
  return bindings.map(binding => binding.name).sort()
}

function assertExact(label: string, actualNames: string[], expectedNames: string[]): void {
  const actual = [...new Set(actualNames)].sort()
  const expected = [...new Set(expectedNames)].sort()
  const actualSet = new Set(actual)
  const expectedSet = new Set(expected)
  const missing = expected.filter(name => !actualSet.has(name))
  const extra = actual.filter(name => !expectedSet.has(name))
  if (actual.length !== EXPECTED || missing.length || extra.length) {
    throw new Error(`${label} parity failed: count=${actual.length}, missing=${missing.join(',')}, extra=${extra.join(',')}`)
  }
  console.log(`${label}: ${actual.length}/${EXPECTED}`)
}

function assertBindings(
  label: string,
  actual: ReadonlyArray<{ name: string; category: string; methodName: string }>,
  expected: ReadonlyArray<{ name: string; category: string; methodName: string }>,
): void {
  assertExact(label, namesFromBindings(actual), expected.map(binding => binding.name))
  const expectedByName = new Map(expected.map(binding => [binding.name, `${binding.category}.${binding.methodName}`]))
  const drift = actual
    .filter(binding => expectedByName.get(binding.name) !== `${binding.category}.${binding.methodName}`)
    .map(binding => binding.name)
  if (drift.length) throw new Error(`${label} namespace/method drift: ${drift.join(',')}`)
}

async function main(): Promise<void> {
  const root = join(import.meta.dirname, '..')
  const manifest = JSON.parse(await readFile(join(root, 'contracts/mcp.tools.json'), 'utf8')) as {
    toolCount: number
    tools: Array<{ name: string; category: string; methodName: string; outputSchema?: Record<string, unknown> }>
  }
  if (manifest.toolCount !== EXPECTED || manifest.tools.length !== EXPECTED) {
    throw new Error(`Manifest count is ${manifest.tools.length}, expected ${EXPECTED}`)
  }
  const missingOutputSchemas = manifest.tools.filter(tool => tool.outputSchema?.type !== 'object').map(tool => tool.name)
  if (missingOutputSchemas.length) {
    throw new Error(`Manifest tools missing output schemas: ${missingOutputSchemas.join(',')}`)
  }
  const expectedNames = manifest.tools.map(tool => tool.name).sort()
  assertBindings('Node memory package', memorySdk.MCP_TOOL_BINDINGS, manifest.tools)
  assertBindings('Node scraper package', scraperSdk.MCP_TOOL_BINDINGS, manifest.tools)
  assertExact('CLI catalog', MCP_TOOL_CATALOG.map(tool => tool.name), expectedNames)
  const expectedCategories = new Map(manifest.tools.map(tool => [tool.name, tool.category]))
  const cliCategoryDrift = MCP_TOOL_CATALOG
    .filter(tool => expectedCategories.get(tool.name) !== tool.category)
    .map(tool => tool.name)
  if (cliCategoryDrift.length) throw new Error(`CLI category drift: ${cliCategoryDrift.join(',')}`)

  const curlDoc = await readFile(join(root, 'docs/curl-tools.md'), 'utf8')
  const curlNames = [...curlDoc.matchAll(/^- `([^`]+)` —/gm)].map(match => match[1])
  assertExact('cURL catalog', curlNames, expectedNames)
}

main().catch(error => {
  console.error(error)
  process.exitCode = 1
})
